### PR TITLE
Add private functions to extensions

### LINKED ISSUE #2265: Extensions Editor - Private function for Extensions
## Description

To avoid code duplication, it is convenient to break them into functions, in the extension I also want to do this, but make some functions private/service, so that the user does not have access to them, but in the extension they can be used.

## Solution suggested

If we talk about implementation, then this is the visibility area (or namespace).
And it has already been implemented, because the life cycle methods are not visible in the event list, for others it is possible to set such a “sign” by method.

### PR DESCRIPTION
**Pull request by arthuro555**
Closes #2265
Adds private functions, functions only available in the current extension.

### CODE REVIEW

File: Core/GDCore/Project/EventsFunction.h [hunk 1]
```diff
@@ -116,6 +117,24 @@ class GD_CORE_API EventsFunction {
    */
   FunctionType GetFunctionType() const { return functionType; };
 
+  /**
+   * \brief Returns true if the function is private.
+   */
+  bool IsPrivate() {
+    if (isPrivate)
```
Comment (arthuro555) [inline Core/GDCore/Project/EventsFunction.h:124]: I probably did a c++ newbie mistake again, but this is the only way I got this to work. `return isPrivate;` always returned true no matter what for some reason.

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 2]
```diff
@@ -86,6 +89,11 @@ export default class EventsFunctionsList extends React.Component<Props, State> {
     searchText: '',
   };
 
+  _togglePrivate = eventsFunction => {
+    eventsFunction.togglePrivate();
+    this.forceUpdate();
```
Comment (arthuro555) [inline newIDE/app/src/EventsFunctionsList/index.js:94]: Not sure if this is necessary or the way to go

File: newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js [hunk 3]
```diff
@@ -37,27 +43,20 @@ export type EnumeratedInstructionOrExpressionMetadata =
   | EnumeratedInstructionMetadata
   | EnumeratedExpressionMetadata;
 
-export const filterEnumeratedInstructionOrExpressionMetadataByScope = <
-  +T: EnumeratedInstructionOrExpressionMetadata
```
Comment (arthuro555) [inline newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js:41]: I'm pretty sure this broke flow but CRA wouldn't compile this after I touched the file.

File: Core/GDCore/Project/EventsFunction.h [hunk 4]
```diff
@@ -116,6 +117,24 @@ class GD_CORE_API EventsFunction {
...
+      return true;
+    else
+      return false;
+  }
+
+  /**
+   * \brief Toggles the privateness of the function.
+   */
+  EventsFunction& TogglePrivate() {
```
Comment (4ian) [inline Core/GDCore/Project/EventsFunction.h:133]: Nitpicking: In general avoid designing classes with `ToggleSomeState`: prefer the more explicit `SetSomeState`.
The reason is because most of the time your client/user will want to set something to a specific value. For example, if I click on a button to make my function private, my goal is not to "toggle the privateness", but to set it to true. If I want to make something public, I don't care about toggling anything, I want to set it to false (and if it was false before, well I still want it false, not true :)).

It's also true in the case of UI framework where some update could be batched: the update should contain the definitive value you want to apply, not a command to toggle - which could be the contrary of what you wish if for some reason the value was inverted in the meantime.

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 5]
```diff
@@ -39,6 +39,9 @@ export type EventsFunctionCreationParameters = {|
 const getEventsFunctionName = (eventsFunction: gdEventsFunction) =>
   eventsFunction.getName();
 
+const getEventsFunctionPrivateness = (eventsFunction: gdEventsFunction) =>
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/index.js:42]: isEventsFunctionPrivate

File: newIDE/app/src/EventsFunctionsList/index.js (hunk 2 above)
Comment (4ian) [reply newIDE/app/src/EventsFunctionsList/index.js:94]: You're changing the value of something that is stored in libGD.js, in a memory that is not understood by React as a "state". So you need to call this.forceUpdate to tell React "hey I know I've not modified a state or a prop, but something still changed and you need to render again this component".

File: newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js (hunk 3 above)
Comment (4ian) [reply newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js:41]: What was the error? Can you try to put back the original typing and paste the error here?

File: newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js [hunk 6]
```diff
@@ -66,10 +65,16 @@ export const filterEnumeratedInstructionOrExpressionMetadataByScope = <
...
       ) === behaviorMetadata.getName()
-    ) {
+    )
+      return true;
+
+    // Show private non-behavior functions when editing the extension
+    if (
+      !behaviorMetadata &&
+      eventsFunctionsExtension &&
+      eventsFunctionsExtension.getName() === type.split(separator)[0]
```
Comment (4ian) [inline newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js:75]: Why is this line needed?

File: Core/GDCore/Project/EventsFunction.h (hunk 1 above)
Comment (4ian) [reply Core/GDCore/Project/EventsFunction.h:124]: Weird, it's very unlikely for the compiler to make a miscompilation of something that simple. Can you try going back to `return isPrivate;`?

**Review by 4ian — COMMENTED** [review]
Thanks for the PR, looks pretty nice :)

File: newIDE/app/src/EventsFunctionsList/index.js [hunk 7]
```diff
@@ -323,6 +337,7 @@ export default class EventsFunctionsList extends React.Component<Props, State> {
                       i18n
                     )}
                     reactDndType="GD_EVENTS_FUNCTION"
+                    isItemBold={getEventsFunctionPrivateness}
```
Comment (4ian) [inline newIDE/app/src/EventsFunctionsList/index.js:340]: Do you think we could just add ` (private)` at the end of the name shown in the list?

File: newIDE/app/src/EventsFunctionsList/index.js (hunk 2 above)
Comment (arthuro555) [reply newIDE/app/src/EventsFunctionsList/index.js:94]: Ok, I was unsure because some component seem to ask the component above via a prop to forceUpdate or something like this.

File: newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js (hunk 6 above)
Comment (arthuro555) [reply newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js:75]: To check if the eventsFunctionExtension is the one the instruction or expression is from?

File: newIDE/app/src/EventsFunctionsList/index.js (hunk 7 above)
Comment (arthuro555) [reply newIDE/app/src/EventsFunctionsList/index.js:340]: Personally, I prefer the boldness, but if you think this is better then yes no problem

File: newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js (hunk 3 above)
Comment (arthuro555) [reply newIDE/app/src/InstructionOrExpression/EnumeratedInstructionOrExpressionMetadata.js:41]: Huh the error seems to be gone now. Weird.

File: Core/GDCore/Project/EventsFunction.h (hunk 1 above)
Comment (arthuro555) [reply Core/GDCore/Project/EventsFunction.h:124]: My computer must have had some sort of problem, it seems I had a lot of very weird issues that just disappeared. This works now.

**Comment by 4ian** [comment]
There is a bug, behavior method marked private are still shown. I'm fixing this.

**Comment by 4ian** [comment]
Seems like the InstructionOrObjectSelector is not "scoped", so is currently showing all instructions/free instructions without filtering. I'm also fixing this.

**Comment by 4ian** [comment]
I've done two fixes and a change:
- InstructionOrObjectSelector was displaying (in the list of "free" instructions or when searching) all instructions without filtering the private ones. This is because `filterEnumeratedInstructionOrExpressionMetadataByScope` was never called (because it was useless before). This is fixed, and the "scope" is passed (so the filtering works according to if you're in a layout or an extension or a behavior).
- setPrivate was never called for behavior methods.
- I've also changed the test to check if the extension is the same as the events function extension of the current scope. For this, I added the `extension` in the scope of an instruction and extension. I can then use this to compare its name with the edited events function extension (if any), rather than infering it from the type. Slightly better, and more consistent with the rest.

Finally, a test was failing that I fixed because you can't directly use `gd` outside of a function - it may not be initialized yet.

Should be good now, waiting for the tests to pass :)
